src/ripple/app/tx/impl/details/NFTokenUtils.cpp

```diff
@@ -520,34 +520,55 @@ findTokenAndPage(
     }
     return std::nullopt;
 }
-void
-removeAllTokenOffers(ApplyView& view, Keylet const& directory)
-{
-    view.dirDelete(directory, [&view](uint256 const& id) {
-        auto offer = view.peek(Keylet{ltNFTOKEN_OFFER, id});
 
-        if (!offer)
-            Throw<std::runtime_error>(
-                "Offer " + to_string(id) + " not found in ledger!");
+int
+removeTokenOffersWithLimit(
+    ApplyView& view,
+    Keylet const& directory,
+    int maxDeletableOffers)
+{
+    if (maxDeletableOffers == 0)
+        return 0;
 
-        auto const owner = (*offer)[sfOwner];
+    std::optional<std::uint64_t> pageIndex{0};
+    int deletedOffersCount = 0;
 
-        if (!view.dirRemove(
-                keylet::ownerDir(owner),
-                (*offer)[sfOwnerNode],
-                offer->key(),
-                false))
-            Throw<std::runtime_error>(
-                "Offer " + to_string(id) + " not found in owner directory!");
+    do
+    {
+        auto const page = view.peek(keylet::page(directory, *pageIndex));
+        if (!page)
+            break;
+
+        // We get the index of the next page in case the current
+        // page is deleted after all of its entries have been removed
+        pageIndex = (*page)[~sfIndexNext];
+
+        auto offerIndexes = page->getFieldV256(sfIndexes);
+
+        // We reverse-iterate the offer directory page to delete all entries.
+        // Deleting an entry in a NFTokenOffer directory page won't cause
+        // entries from other pages to move to the current, so, it is safe to
+        // delete entries one by one in the page. It is required to iterate
+        // backwards to handle iterator invalidation for vector, as we are
+        // deleting during iteration.
+        for (int i = offerIndexes.size() - 1; i >= 0; --i)
```

Comment:
Nitpick. Depending on the scenario, we could have an underflow here.

Also, it seems that we can use a range-based loop here.

Reply:
Looping by index backwards was chosen on purpose to delete the entries from `STVector256`

Reply:
Since we are using C++20, we could use ```std::ranges::view``` for looping backward

`for (auto const& offerIndex :  offerIndexes | std::views::reverse) {...}`

Depending on whether the underlying data type is primitive or complex, we can use `auto const` or `auto const&`.

Reply:
hmm...it says that `views` is is not in namespace `std`. I think it is because apple clang does not support views, and im on mac

Reply:
You may need to `#include <ranges>`.  It's also certainly possible that the clang library does not have all of the C++20 features yet.  We've had that problem with other C++20 features.

Reply:
Tried that and still won't work. And ranges isn't supported by apple clang either (red box)
![image](https://user-images.githubusercontent.com/35279399/215866789-28caa935-5c61-4f9c-a140-96d75fd67766.png)

Reply:
Saw on [stackoverflow](https://stackoverflow.com/questions/73628848/is-the-stdviews-namespace-not-available-in-xcodes-c?rq=1)  that `views` isn't in apple clang

Reply:
But I think i can just keep index looping as it is, shouldnt be a problem